lib/Backends/CPU/CPUFunction.h

```diff
@@ -27,12 +27,29 @@ class CPUFunction final : public CompiledFunction {
   /// The LLVM JIT engine. The jit must be initialized after the ctor
   /// initializes the LLVM backends.
   std::unique_ptr<llvm::orc::GlowJIT> JIT_;
-  /// This represents the heap, that stores the activations at runtime.
-  void *heap_;
+
+  /// runtimeInfo contains symbol offsets and allocation sizes
+  std::unique_ptr<runtimeInfo> runtimeInfo_;
+
+  uint8_t *baseConstantWeightVarsAddress_;
+  uint8_t *baseActivationsAddress_;
+  uint8_t *baseMutableWeightVarsAddress_;
 
 public:
   /// Ctor.
-  CPUFunction(std::unique_ptr<llvm::orc::GlowJIT> JIT, void *heap);
+  CPUFunction(std::unique_ptr<llvm::orc::GlowJIT> JIT,
+              std::unique_ptr<runtimeInfo> runtimeInfo);
+
+  /// Copy Function and Constant Memory to device.
+  void loadFunctionOnDevice();
```

Comment:
I wonder if we always want to combine copying of a function (on a CPU no code needs to be copied, but on accelerators we may need to actually copy code) and copying of constant weights into a single function like `loadFunctionOnDevice`? May be it should be two functions? Or a single function that internally calls two different functions?

Reply:
A more general question: Should the CompiledFunction class be responsible for copying to/from devices and for memory allocation? It seems to me that these are runtime/execution aspects and thus belong into a separate API. CompiledFunction objects themselves are passive objects that should be manipulated/executed by other subsystems IMHO.

Reply:
> Compiled function already has execute which is responsible for the actual execution, so I can see a case for keeping all runtime logic co-located. But maybe it's cleaner to separate that functionality to a separate API. Can we remove execute too? Then CompiledFunction would be just the code itself.

Yeah, I think mid term we most likely want the CompiledFunction to be just the code itself and have a separate API/logical system components to run/execute one or more CompiledFunctions. In case of a CPU execution, such a component may be e.g. responsible for managing threads, deciding which CompiledFunction should run in which thread, handle thread local memory @rdzhabarov mentioned, etc.